src/System.IO.FileSystem/src/System/IO/Win32FileStreamAsyncResult.cs

```diff
@@ -96,232 +53,147 @@ unsafe private sealed class FileStreamAsyncResult : IAsyncResult
-                // call EndRead or EndWrite.  There's no reason to use an async 
-                // delegate here - we're already on a threadpool thread.  
-                // IAsyncResult's completedSynchronously property must return
-                // false here, saying the user callback was called on another thread.
-                asyncResult._completedSynchronously = false;
-                asyncResult._isComplete = true;
-
-                Interlocked.MemoryBarrier();
-
-                // ensure _isComplete is set before reading _waitHandle
-                // The OS does not signal this event.  We must do it ourselves.
-                ManualResetEvent wh = asyncResult._waitHandle;
-                if (wh != null)
+                if (errorCode != 0 && errorCode != Win32FileStream.ERROR_BROKEN_PIPE && errorCode != Win32FileStream.ERROR_NO_DATA)
                 {
-                    Debug.Assert(!wh.GetSafeWaitHandle().IsClosed, "ManualResetEvent already closed!");
-                    bool r = wh.Set();
-                    Debug.Assert(r, "ManualResetEvent::Set failed!");
-                    if (!r) throw Win32Marshal.GetExceptionForLastWin32Error();
+                    overlappedToTask._completionSource.TrySetException(Win32Marshal.GetExceptionForWin32Error(unchecked((int)errorCode)));
+                }
+                else
+                {
+                    overlappedToTask._completionSource.TrySetResult((int)numBytes + overlappedToTask._numBufferedBytes);
                 }
-
-                AsyncCallback userCallback = asyncResult._userCallback;
-                if (userCallback != null)
-                    userCallback(asyncResult);
             }
 
-            internal void Cancel()
+            private static void Cancel(object state)
             {
-                Debug.Assert(_handle != null, "_handle should not be null.");
-                Debug.Assert(_overlapped != null, "Cancel should only be called on true asynchronous FileStreamAsyncResult, i.e. _overlapped is not null");
-
-                if (IsCompleted)
-                    return;
+                // WARNING: This may potentially be called under a lock (during cancellation registration)
 
-                if (_handle.IsInvalid)
-                    return;
+                FileStreamAsyncResult asyncResult = state as FileStreamAsyncResult;
+                Debug.Assert(asyncResult != null, "Unknown state passed to cancellation");
+                Debug.Assert(asyncResult._overlapped != null && !asyncResult._completionSource.Task.IsCompleted, "IO should not have completed yet");
 
-                bool r = Interop.mincore.CancelIoEx(_handle, _overlapped);
-                if (!r)
+                // If the handle is still valid, attempt to cancel the IO
+                if ((!asyncResult._handle.IsInvalid) && (!Interop.mincore.CancelIoEx(asyncResult._handle, asyncResult._overlapped)))
```

Comment:
It's a fair point. Additionally - if I switch FileStreamAsyncResult to inherit from TaskCompletionSource, then I would be locking on "this" (which I really don't want to do).
So... looks like I need to switch to the lock-free version...